Closes #506: Enabling custom linters, adding preference for {cli} fns over `stop()` and `abort()`

manciniedoardo opened the pull request:
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/dev/CONTRIBUTING.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiraldev/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

bundfussr commented:
I think `source_file` doesn't contain the complete .R file, e.g., `lintr.R` but you get separate `source_file` objects for each "block" in `lintr.R`, i.e., each comment and each function definition.

manciniedoardo commented:
I tested on admiralophtha, and it's super easy to use these global admiral linters:
1. Rename `.lintr` to `.lintr.R`
2. Replace contents of file with:
```
library(admiraldev)

linters <- admiraldev:::admiral_linters()

exclusions <- list(
  "R/data.R" = Inf,
  "inst" = list(undesirable_function_linter = Inf),
  "vignettes" = list(undesirable_function_linter = Inf)
)
```
3. Profit

manciniedoardo commented on R/linters.R:
```diff
@@ -0,0 +1,29 @@
+      cyclocomp_linter(complexity_limit = 20),
+      undesirable_function_linter = undesirable_function_linter(
+        fun = c(
+          "default" = default_undesirable_functions,
```
You're right, this syntax was incorrect, thanks for spotting.

manciniedoardo commented on R/linters.R:
```diff
@@ -0,0 +1,29 @@
+        fun = c(
+          "default" = default_undesirable_functions,
+          "stop" = "please use cli_abort() from the {cli} package",
+          "warning" = "please use cli_warning() from the {cli} package"
```
Good idea. I'll add those two for now - though i think they may fire for admiraldev!!

bundfussr replied:
We should add `message` as well.

manciniedoardo replied:
added

manciniedoardo commented on R/linters.R:
On second thoughts though, you could argue that if for some reason an ext package wanted to add another linter, that is not used elsewhere they could make an edit to `admiral_linters()` to add it themselves as an option that is preset to NULL but overwrite it in their package

bundfussr replied:
Maybe add the `...` argument to `admiral_linters()` for specifying any arguments for `linters_with_defaults()`?

We then need to use our defaults only for linters which are not specified via `...`.

bundfussr replied:
Looks good to me.

bms63 commented on R/linters.R:
okay we have a lot of updating to do then - the lintr check is long!

manciniedoardo commented:
> Should we include a little section in the programming strategy and the use of lintrs and what we are trying to achieve?

I had already added this - are you proposing to expand?

<img width="734" height="326" alt="image" src="https://github.com/user-attachments/assets/c51a74c5-9e99-45ba-9ff7-e02f67d986ba" />

manciniedoardo commented:
@bms63 @bms63 if you're both happy with this approach, I can make the required changes (to get our own checks to pass!) and then open up to the rest of the team at Monday's meet to see if there's anything else we want to add. Then we can start implementing on admiral and ext pkgs?

bundfussr commented:
> @bms63 @bms63 if you're both happy with this approach, I can make the required changes (to get our own checks to pass!) and then open up to the rest of the team at Monday's meet to see if there's anything else we want to add. Then we can start implementing on admiral and ext pkgs?

Looks good to me.

manciniedoardo commented:
@bms63 @bundfussr 

As discussed, I've removed the lintr dependency. Instead of pharmaverse4devs/add-in I just moved the custom linter function to `inst/lintr/linters/R`. Then any other package wishing to use it can just have the following `.lintr.R` config file:
```
library(lintr)

source(system.file("lintr/linters.R", package = "admiraldev"))

linters <- admiral_linters() #Any exclusions or changes can be manually specified here

exclusions <- list(
  "R/data.R" = Inf,
  "inst" = list(undesirable_function_linter = Inf),
  "R/assertions.R" = list(undesirable_function_linter = Inf),
  "vignettes" = list(undesirable_function_linter = Inf)
)
```
What do you think? If happy please approve and merge and I can help this trickle down to the extensions, as well as admiral. For the latter I expect quite a lot of changes required to get the initial linter check to pass our own requirements, but I haven't verified that yet.

bundfussr commented:
> @bms63 @bundfussr this is now ready for final review before opening up to the rest of the team and then merging. I had to ignore `R/assertions.R` for lintr checks because we use `message` as an argument way too many times and gives a false positive. Elsewhere, I renamed to `msg` instead. For the other lints I replaced with the suggested functions or ignored with `#nolint` if just in a test file.

I think we should use `symbol_is_undesirable = FALSE` for the `undesirable_function_linter`. We have this in admiral. It should avoid that the symbol `message` is reported as an issue.

bundfussr commented on inst/lintr/linters.R:
```diff
@@ -0,0 +1,61 @@
+        "message" = "please use cli_inform() from the {cli} package",
+        "ifelse" = "please use vectorized if_else() from the {dplyr} package",
+        "grepl" = "please use str_detect() from the {stringr} package"
+      )
```
```suggestion
      ),
      symbol_is_undesirable = TRUE
```
This should allow us to use the `undesirable_function_linter` in `assertions.R`.

bundfussr commented on .lintr.R:
```diff
@@ -0,0 +1,12 @@
+exclusions <- list(
+  "R/data.R" = Inf,
+  "inst" = list(undesirable_function_linter = Inf),
+  "R/assertions.R" = list(undesirable_function_linter = Inf),
```
Could be removed if we use `symbol_is_undesirable = TRUE`.

bundfussr replied:
Sorry, I meant `symbol_is_undesirable = FALSE`.

manciniedoardo replied:
ah yes thanks, fixed.

bundfussr commented on vignettes/programming_strategy.Rmd:
```diff
@@ -232,11 +232,18 @@ Arguments which only expect one value or variable name must be a singular versio
+
+With regards to [lintr](https://github.com/jimhester/lintr), `{admiral}` and all
+related packages should maintain consistent linting standards by ensuring that their 
+`.lintr` configuration files use the `admiral_linters()` function. This contains 
```
```suggestion
`.lintr.R` configuration files use the `admiral_linters()` function. This contains 
```

bundfussr commented on vignettes/programming_strategy.Rmd:
```diff
@@ -232,11 +232,18 @@ Arguments which only expect one value or variable name must be a singular versio
+agreed-upon preferences and conventions such as avoiding the use of `stop()` and
+`warning()` in favor of `cli::abort()` and `cli::warn()`.
+The `admiral_linters()` function is stored under `inst/lintr/linters.R` in `{admiraldev}` 
+(so as not to expose it to users) and can be loaded within the `.lintr` configuration 
```
```suggestion
(so as not to expose it to users) and can be loaded within the `.lintr.R` configuration 
```

bundfussr commented on vignettes/programming_strategy.Rmd:
```diff
@@ -232,11 +232,18 @@ Arguments which only expect one value or variable name must be a singular versio
+The `admiral_linters()` function is stored under `inst/lintr/linters.R` in `{admiraldev}` 
+(so as not to expose it to users) and can be loaded within the `.lintr` configuration 
+file with `source(system.file("lintr/linters.R", package = "admiraldev")`. 
 
```
Should we display an example `.lintr.R` file here?

manciniedoardo commented on vignettes/programming_strategy.Rmd:
```diff
@@ -232,11 +232,47 @@ Arguments which only expect one value or variable name must be a singular versio
+exclusions <- list(
+  "R/data.R" = Inf,
+  "inst" = list(undesirable_function_linter = Inf),
+  "R/assertions.R" = list(undesirable_function_linter = Inf),
```
good point, removed
